ПР8

Megrief opened the pull request:
- Все объекты распределены по пакетам.
- Взаимодействие между слоем ui и слоем data происходит через слой domain посредством интерфейсов.
- Вся логика плеера вынесена из AudioplayerActivity. Взаимодействие с плеером осуществляется через PlayerInteractor.

Megrief committed:
Рефакторинг:
- Все объекты распределены по пакетам.
- Взаимодействие между слоем ui и слоем data происходит через слой domain посредством интерфейсов.
- Вся логика плеера вынесена из AudioplayerActivity. Взаимодействие с плеером осуществляется через PlayerInteractor.

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/data/network/NetworkClient.kt:
```diff
@@ -0,0 +1,5 @@
+package com.example.playlistmaker.data.network
+
+interface NetworkClient {
+    fun doRequest(dto: Any): Response
```
Привет! Здесь рекомендую сделать функцию `search` или `doSearch` с аргументом типа `TrackSearchRequest`. Чтобы конкретный запрос работал с конкретным реквестом. Это позволит тебе избавиться от обработки лишних условий и сделает этот метод менее абстрактным.

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/data/network/RetrofitClientImpl.kt:
```diff
@@ -0,0 +1,31 @@
+            }
+            val body = response?.body() ?: Response()
+
+            body.apply { resultCode = response?.code() ?: 400 }
```
использовать числа в коде является плохой практикой, которая называется Магическое число. https://ru.wikipedia.org/wiki/%D0%9C%D0%B0%D0%B3%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B5_%D1%87%D0%B8%D1%81%D0%BB%D0%BE_(%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)#%D0%9F%D0%BB%D0%BE%D1%85%D0%B0%D1%8F_%D0%BF%D1%80%D0%B0%D0%BA%D1%82%D0%B8%D0%BA%D0%B0_%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F можно почитать здесь подробнее. Рекомендую вынести значение кода в константу

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/data/repo_impl/SearchRepoImpl.kt:
```diff
@@ -0,0 +1,40 @@
+    override fun search(term: String): List<Track>? {
+        val response = networkClient.doRequest(TrackSearchRequest(term))
+
+        return if (response.resultCode == 200) {
```
здесь тоже значение кода ответа лучше вынести в константу

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/data/repo_impl/SearchRepoImpl.kt:
```diff
@@ -0,0 +1,40 @@
+                    Track(
+                        trackId = it.trackId ?: -1,
+                        trackName = it.trackName ?: "",
+                        trackTime = it.trackTime?.let { len -> getLength(len) } ?: "00:00",
```
`"00:00"` советую вынести в ресурсы, т.к. это плейсхолдер. В целом логика форматирования дат и времени относится к слою презентации, поэтому тебе стоит хранить здесь все так же `Long`, а `getLength(len)` вынести в слой `presentation` и использовать. когда `trackTime` устанавливается в `textView`

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/data/repo_impl/SharedPrefsList.kt:
```diff
@@ -0,0 +1,23 @@
+class SharedPrefsList(context: Context) : StorageManagerRepo<List<Track>> {
+    private val sharedPrefs = context.getSharedPreferences(App.PLAYLIST_MAKER, Context.MODE_PRIVATE)
+    override fun save(key: String, item: List<Track>) {
+        Gson().toJson(TrackListDto(item)).let {
```
можно один раз инициализировать `Gson()` и хранить его в поле класса, так будет затрачено меньше времени и памяти на выполнение этих методов

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/data/repo_impl/SharedPrefsTrack.kt:
```diff
@@ -0,0 +1,23 @@
+    private val sharedPrefs = context.getSharedPreferences(App.PLAYLIST_MAKER, Context.MODE_PRIVATE)
+
+    override fun save(key: String, item: Track?) {
+        Gson().toJson(item).let {
```
здесь тоже `Gson()` можно инициализировать один раз

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/domain/interactors_impl/StorageInteractorTrackList.kt:
```diff
@@ -0,0 +1,26 @@
+class StorageInteractorTrackList(
+    private val repository: StorageManagerRepo<List<Track>>
+) : StorageInteractor<List<Track>> {
+    private val executor = Executors.newCachedThreadPool()
```
здесь нет необходимости в экзекьюторе. `SharedPreferences`, на которых основана работа твоего репозитория, работают асинхронно, если использовать `apply`

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/domain/repositories/StorageManagerRepo.kt:
```diff
@@ -0,0 +1,7 @@
+package com.example.playlistmaker.domain.repositories
+
+interface StorageManagerRepo<T> {
```
неплохое использование generics

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/presentaion/ui/audioplayer/AudioplayerBinder.kt:
```diff
@@ -0,0 +1,50 @@
+
+object AudioplayerBinder {
+
+    fun bind(activity: AudioplayerActivity, track: Track) {
```
не рекомендуется оставлять ссылки на активити в синглтон классах, которым является `object`. Это может привести к утечкам памяти. Советую просто вынести этот код как отдельные приватные функции в классе `AudioplayerActivity`

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/presentaion/ui/audioplayer/PlayerInteractor.kt:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    companion object {
+        private const val DELAY = 333L
```
хорошим тоном является указывать единицы измерения у констант, например DELAY_MILLS

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/presentaion/utils/Player.kt:
```diff
@@ -0,0 +1,56 @@
+        playerState = STATE_PAUSED
+    }
+
+    fun releaseResources() {
```
публичные функции следует располагать выше приватных, согласно соглашениям кода Kotlin

vlad-zotov commented on app/src/main/java/com/example/playlistmaker/presentaion/utils/Player.kt:
```diff
@@ -0,0 +1,56 @@
+
+class Player(
+    url: String,
+    onPrepared: Runnable,
```
здесь тебе лучше подойдут лямбды ` () -> Unit` в качестве возвращаемого результата. `Runnable` это больше вещь из мира многопоточности

Megrief committed:
Исправлены недочеты:
- Убраны executors из StorageInteractor;
- Логика из объекта AudioplayerBinder возвращена в AudioplayerActivity;
- Преобразование времени трека перенесено в слой presentation;
- Исправлена ошибка, при создании экрана AudioplayerActivity;
- Учтены замечания к качеству кода.
